lib/sshutils/sftp/parse.go

```diff
@@ -17,64 +17,119 @@ limitations under the License.
 package sftp
 
 import (
-	"regexp"
+	"strings"
 
 	"github.com/gravitational/trace"
 
 	"github.com/gravitational/teleport/lib/utils"
 )
 
-var reSFTP = regexp.MustCompile(
-	// optional username, note that outside group
-	// is a non-capturing as it includes @ signs we don't want
-	`(?:(?P<username>.+)@)?` +
-		// either some stuff in brackets - [ipv6]
-		// or some stuff without brackets and colons
-		`(?P<host>` +
-		// this says: [stuff in brackets that is not brackets] - loose definition of the IP address
-		`(?:\[[^@\[\]]+\])` +
-		// or
-		`|` +
-		// some stuff without brackets or colons to make sure the OR condition
-		// is not ambiguous
-		`(?:[^@\[\:\]]+)` +
-		`)` +
-		// after colon, there is a path that could consist technically of
-		// any char including empty which stands for the implicit home directory
-		`:(?P<path>.*)`,
-)
-
-// Destination is SCP destination to copy to or from
+// Destination is a remote SFTP destination to copy to or from.
 type Destination struct {
 	// Login is an optional login username
 	Login string
 	// Host is a host to copy to/from
-	Host utils.NetAddr
+	Host *utils.NetAddr
 	// Path is a path to copy to/from.
 	// An empty path name is valid, and it refers to the user's default directory (usually
 	// the user's home directory).
 	// See https://tools.ietf.org/html/draft-ietf-secsh-filexfer-09#page-14, 'File Names'
 	Path string
 }
 
-// ParseSCPDestination takes a string representing a remote resource for SFTP
-// to download/upload, like "user@host:/path/to/resource.txt" and parses it into
+// ParseDestination takes a string representing a remote resource for SFTP
+// to download/upload in the form "[user@]host:[path]" and parses it into
 // a structured form.
 //
 // See https://tools.ietf.org/html/draft-ietf-secsh-filexfer-09#page-14, 'File Names'
 // section about details on file names.
-func ParseDestination(s string) (*Destination, error) {
-	out := reSFTP.FindStringSubmatch(s)
-	if len(out) < 4 {
-		return nil, trace.BadParameter("failed to parse %q, try form user@host:/path", s)
+func ParseDestination(input string) (*Destination, error) {
+	firstColonIdx := strings.Index(input, ":")
```

Comment:
What if the host being provided is an IPV6 address, there will be many colons.  I believe the standard way to provide an IPV6 address to scp is to surround it in `[ff:ff:...]`

To fix this perhaps we need to be looking for the last index, then confirm that index is not surrounded by `[` `]`?

Ironically enough we also have a semi-relevant issue for that: https://github.com/gravitational/teleport/issues/2753

Reply:
I didn't think of the possibility of the path containing an IPv6 address, so I added a few test cases and they passed without needing to modify anything: https://github.com/gravitational/teleport/blob/cd1e012d9df7efe712528c9aea8173e862d05f0b/lib/sshutils/sftp/parse_test.go#L178-L213

Reply:
@capnspacehook I checked out the branch so I could run the code, and I now realize what's going on.  It is functioning correctly, but it is quite confusing.

Take your existing example test here: https://github.com/gravitational/teleport/blob/cd1e012d9df7efe712528c9aea8173e862d05f0b/lib/sshutils/sftp/parse_test.go#L61-L71

With the input of `[::1]:/etc/nginx.co:nf` line `47` will parse the first index to be `1`.  This will satisfy the `path` check directly below, which is quite confusing because it at that point has done nothing to verify the file / path exists.  Thus it's quite misleading when an ipv6 address is provided as the host.

I believed originally this set of a chain of parsing errors when reviewing the code, however when running it locally I now realize that this error condition is corrected once the IPv6 host is parsed (path start index corrected at line 86, and empty path recognized at line `126`).

My mistake in identifying this issue, but I think it's worth recognizing that this is certainly very confusing to follow.  In any case now that I see it's functioning correctly I will approve the PR.  Thank you for helping explore this with me!

Reply:
No problem, thanks for the feedback! I added a few more comments, but is there anything else I can do to make the code less confusing in your opinion?

Reply:
> is there anything else I can do to make the code less confusing in your opinion?

This would be my only recommendation:

The host parsing logic seems pretty pivotal, if we could detect if it's an IPv6 before attempting to do other parsing or validation it may be easier to read.  This could mean that we even re-parse some parts of the input after the host has been determined in order to do that validation.  But it would fix the issue of having some of the validation not be accurate to what it says it's verifying

That would make it personally easier for me to understand, but it also may overall add more code.  
It always seems cleaner conceptually before you actually try to implement it.  So I entirely defer to you if you think it's an improvement or not.

Reply:
Hmm, yeah I think that would complicate that code a bit more, so I think I'll leave it as is. Thanks again for your feedback, I appreciate it!